### DISCUSSION TITLE
Save/Cancel buttons for the toolbar actions

### DISCUSSION THREAD

**Discussion by gagamilano** [geoman-io/leaflet-geoman#1116] [Questions & Help]
I would like the user to confirm or cancel the edits he's making on the geometries, like in this Leaflet.Draw demo: https://leaflet.github.io/Leaflet.draw/docs/examples/full.html
So I'd like to add Save and Cancel buttons to the following actions: Edit, Drag, Cut, Rotate. I reckon the Save's button is just the "finishMode" toolbar action, so that should be easy. The Cancel however is proving to be a little bit harder to code in a clean way.

This is what I have so far...

Setup Geoman normally, then replace the various button's actions:
```
const getButtonActions = btnName => {
    return [
        "finishMode",
        {
            text: "Annulla",
            onClick: () => {
                if (this._coords && this.target) {
                    this.target.setLatLngs(this._coords);
                    this._coords = null;
                }

                geoman.Toolbar.buttons[btnName]._triggerClick();
            }
        }
    ]
}
geoman.Toolbar.changeActionsOfControl("Edit", getButtonActions("editMode"));
geoman.Toolbar.changeActionsOfControl("Drag", getButtonActions("dragMode"));
geoman.Toolbar.changeActionsOfControl("Rotate", getButtonActions("rotateMode"));
```

Then inside the various pm:edit, pm:dragstart etc events I memorize the original coordinates inside the this._coords object. In my application only one layer max has geoman enabled at any given time, and this layer will be in the this.target object. So there's no need to handle possible multiple rollbacks.

Is there a better way of doing this?

**Comment by cubecleveland** [comment]
Not sure if this will work for you but this is how i handled this. i saves a 'state' inside the layer itself on each edit in an array.
layer._state = layer.geltLanlngs.push() -- > whatever u need there
You can also save the pointer to the state like this layer_pointerToCurrentState

Tehn i just went back to step inside this array on every "undo" action.
layer.setLatLngs(_player_state(pointerToState) -->  Yu get the idea...

Not sure if this is the "best" way to handle it.. but i thought i throw my own 25 cents.

You should probably wait to hear from the orginal contributor as his advice are most valuable if this doesnt work for you 
:)

**Reply by gagamilano** [reply]
I edited my post literally 20 seconds after you posted this and we came to a similar conclusion. My problem's easier though because only one layer can have geoman enabled on it at a time in my application, so I'm never gonna have to rollback more than 1 shape.
